omp/test/solver/bicg_kernels.cpp

```diff
@@ -0,0 +1,303 @@
+
+    std::unique_ptr<Mtx> d_b;
+    std::unique_ptr<Mtx> d_r;
+    std::unique_ptr<Mtx> d_z;
+    std::unique_ptr<Mtx> d_p;
+    std::unique_ptr<Mtx> d_q;
+    std::unique_ptr<Mtx> d_r2;
+    std::unique_ptr<Mtx> d_z2;
+    std::unique_ptr<Mtx> d_p2;
+    std::unique_ptr<Mtx> d_q2;
+    std::unique_ptr<Mtx> d_x;
+    std::unique_ptr<Mtx> d_beta;
+    std::unique_ptr<Mtx> d_prev_rho;
+    std::unique_ptr<Mtx> d_rho;
+    std::unique_ptr<gko::Array<gko::stopping_status>> d_stop_status;
+};
+
+
+TEST_F(Bicg, OmpBicgInitializeIsEquivalentToRef)
+{
+    initialize_data();
+
+    gko::kernels::reference::bicg::initialize(
+        ref, b.get(), r.get(), z.get(), p.get(), q.get(), prev_rho.get(),
+        rho.get(), r2.get(), z2.get(), p2.get(), q2.get(), stop_status.get());
+    gko::kernels::omp::bicg::initialize(
+        omp, d_b.get(), d_r.get(), d_z.get(), d_p.get(), d_q.get(),
+        d_prev_rho.get(), d_rho.get(), d_r2.get(), d_z2.get(), d_p2.get(),
+        d_q2.get(), d_stop_status.get());
+
+    GKO_ASSERT_MTX_NEAR(d_r, r, 1e-14);
+    GKO_ASSERT_MTX_NEAR(d_z, z, 1e-14);
+    GKO_ASSERT_MTX_NEAR(d_p, p, 1e-14);
+    GKO_ASSERT_MTX_NEAR(d_q, q, 1e-14);
+    GKO_ASSERT_MTX_NEAR(d_r2, r2, 1e-14);
+    GKO_ASSERT_MTX_NEAR(d_z2, z2, 1e-14);
+    GKO_ASSERT_MTX_NEAR(d_p2, p2, 1e-14);
+    GKO_ASSERT_MTX_NEAR(d_q2, q2, 1e-14);
+    GKO_ASSERT_MTX_NEAR(d_prev_rho, prev_rho, 1e-14);
+    GKO_ASSERT_MTX_NEAR(d_rho, rho, 1e-14);
+    GKO_ASSERT_ARRAY_EQ(d_stop_status, stop_status);
+}
+
+
+TEST_F(Bicg, OmpBicgStep1IsEquivalentToRef)
+{
+    initialize_data();
+
+    gko::kernels::reference::bicg::step_1(ref, p.get(), z.get(), p2.get(),
+                                          z2.get(), rho.get(), prev_rho.get(),
+                                          stop_status.get());
+    gko::kernels::omp::bicg::step_1(omp, d_p.get(), d_z.get(), d_p2.get(),
+                                    d_z2.get(), d_rho.get(), d_prev_rho.get(),
+                                    d_stop_status.get());
+
+    GKO_ASSERT_MTX_NEAR(d_p, p, 1e-14);
+    GKO_ASSERT_MTX_NEAR(d_z, z, 1e-14);
+    GKO_ASSERT_MTX_NEAR(d_p2, p2, 1e-14);
+    GKO_ASSERT_MTX_NEAR(d_z2, z2, 1e-14);
+}
+
+
+TEST_F(Bicg, OmpBicgStep2IsEquivalentToRef)
+{
+    initialize_data();
+
+    gko::kernels::reference::bicg::step_2(
+        ref, x.get(), r.get(), r2.get(), p.get(), q.get(), q2.get(), beta.get(),
+        rho.get(), stop_status.get());
+    gko::kernels::omp::bicg::step_2(
+        omp, d_x.get(), d_r.get(), d_r2.get(), d_p.get(), d_q.get(), d_q2.get(),
+        d_beta.get(), d_rho.get(), d_stop_status.get());
+
+    GKO_ASSERT_MTX_NEAR(d_x, x, 1e-14);
+    GKO_ASSERT_MTX_NEAR(d_r, r, 1e-14);
+    GKO_ASSERT_MTX_NEAR(d_r2, r2, 1e-14);
+    GKO_ASSERT_MTX_NEAR(d_p, p, 1e-14);
+    GKO_ASSERT_MTX_NEAR(d_q, q, 1e-14);
+    GKO_ASSERT_MTX_NEAR(d_q2, q2, 1e-14);
+}
+
+
+TEST_F(Bicg, ApplyIsEquivalentToRef)
+{
+    auto mtx = gen_mtx(50, 50);
+    make_spd(mtx.get());
```

Comment:
I don't think you need to make the matrix SPD for `BiCG`? Being non SPD could make the solver less accurate/crash though, so an approach could be to both test SPD and non SPD.

Reply:
I am now using the an1.mtx matrix for the non spd test for cuda and hip.